Userland/Games/Chess/ChessWidget.cpp

```diff
@@ -536,91 +538,191 @@ ErrorOr<String> ChessWidget::get_fen() const
     return TRY(m_playback ? m_board_playback.to_fen() : m_board.to_fen());
 }
 
-ErrorOr<void> ChessWidget::import_pgn(Core::File& file)
+ErrorOr<void, PGNParseError> ChessWidget::import_pgn(Core::File& file)
 {
-    m_board = Chess::Board();
+    reset();
+    enum TokenType {
+        MOVE,
+        TAG_SYMBOL,
+        TAG_STRING,
+        COMMENT,
+        GAME_TERMINATOR,
+        RAV_START,
+        RAV_END,
+        NAG
+    };
 
-    ByteBuffer bytes = TRY(file.read_until_eof());
-    StringView content = bytes;
-    auto lines = content.lines();
-    StringView line;
-    size_t i = 0;
+    struct Token {
+        TokenType type;
+        StringView value;
+    };
+    auto maybe_file = file.read_until_eof();
+    if (maybe_file.is_error()) {
+        return PGNParseError::from_string_formatted(String::formatted("Could not read file"));
+    }
+    ByteString bytes = ByteString { maybe_file.release_value().bytes() };
+    Vector<Token> tokens;
+    Vector<StringView> rav_stack;
+    // FIXME: Engine cannot parse suffixes ? and !.
+    StringView suffix_characters = "+#"sv;
+    StringView closing_characters = "]})"sv;
+    StringView opening_characters = "({["sv;
+    LineTrackingLexer lexer { StringView { bytes } };
+
+    while (!lexer.is_eof()) {
+
+        if (lexer.next_is(is_any_of(closing_characters))) {
+            return PGNParseError::from_string_formatted(String::formatted("Unexpected character: {}.\n(line {} column {})", lexer.consume(1), lexer.current_position().line + 1, lexer.current_position().column));
+
+        } else if (lexer.next_is('[')) {
+            lexer.consume_specific('[');
+            auto value = lexer.consume_until(" ");
+            tokens.append(Token { TokenType::TAG_SYMBOL, value });
+            lexer.ignore_while([](char c) { return is_ascii_space(c); });
```

Comment:
I *think* this line can just be:

```suggestion
            lexer.ignore_while(is_ascii_space);
```

Same for the other uses below.